### ISSUE TITLE
WIP: Propagating informational & error messages from workers to frontend

### ISSUE THREAD

**Issue by lfse-slafleur** [Project-OMOTES/architecture-documentation#38]
# Issue description
The worker expresses multiple types of information that is useful to show to a user of the frontend regarding their submitted job. The following types of information have been gathered:
- If a job succeeds:
  - Information regarding assets. (also called 'Asset Feedback')
  - Information regarding the ESDL as a whole.
- If a job fails:
  - Error messages why the job couldn't be processed.

Currently errors are caught and the logs of a job are forwarded to the frontend. However, this is not a user-friendly way of showing any type of errors. Mostly because it doesn't only show the error but also all kinds of debug information. While this debug information is useful for developers, a user is unable to currently interpret it. Therefore, we should provide a way for user-friendly error messages to be propagated from the worker to the frontend.

Also, in the near future, we expect that the workers are able to supply additional information to a user regarding the output ESDL. This is also referred to as 'Asset Feedback'. Specifically MapEditor already supports a method for showing informational messages tied to a specific asset and MapEditor requires a specific format to be used for these informational messages.

## Severity
The proposal is to use the same levels for severity as the logging module follow:
- DEBUG <-- Useful only for developers. Should not be shown to users per default. Specifically optimizer team expects this level to be useful.
- INFO <-- Shows extra information regarding the output ESDL without questioning the validity of the result.
- WARNING <-- Something regarding the output ESDL is questioning its validity. The user should inspect if the warning is a cause of concern.
- ERROR <-- Job (optimization or simulation) could not succeed because there was something wrong in the input data or the job couldn't be processed due to an IT infrastructure issue (e.g. job timeout, job reached delivery limit). No output ESDL.

## Origin of messages
While the worker will perform the job, the orchestrator may also introduce informational messages. Specifically if a job timeout or delivery limit was reached the orchestrator will provide the necessary error message.

# Solution

## Displaying the informational message(s) at the frontend
Each frontend will want control on how to display the messages. Specifically localization (language in which the message is shown) may be different depending on the frontend and the user

**Comment by edwinmatthijssen** [comment]
I definitely want to extend the possibilities on this topic (handling model feedback, logging, progress reporting, and so on) in the ESDL MapEditor as well, so for asset-feedback there is no work (or hardly any work), but in general this statement is not valid.

**Comment by edwinmatthijssen** [comment]
True....         adding non-asset specific feedback to the MapEditor is easy and good to add!

**Comment by cwang39403** [comment]
I would prefer to forward the messages as they are from a maintenance perspective. 

- This proposal _could_ be a nice addition to the frontend user to navigate when errors occur, but assuming most users using OMOTES are likely to have a certain technical background and probably also going through relevant training already, the additional layer of adding error codes and parsing technical message does not add that much value at this moment in my opinion.
- I am more concerned about the maintenance of this additional layer as this requires inputs from different teams. Now it is still easy to find each other during the active development phase, but it is hard to guarantee in the longer term. Also, any change request could introduce changes in different dependent components.

For the time being, if we constantly receiving similar support requests from the users, I would look into the direction of having a Q&A section on the (architecture) documentation page that lists a few commonly seen technical/error messages to help users with the next steps.

**Comment by lfse-slafleur** [comment]
Hea Cheng-Kai!

> I would prefer to forward the messages as they are from a maintenance perspective.
> 
>     * This proposal _could_ be a nice addition to the frontend user to navigate when errors occur, but assuming most users using OMOTES are likely to have a certain technical background and probably also going through relevant training already, the additional layer of adding error codes and parsing technical message does not add that much value at this moment in my opinion.

From MapEditor perspective it doesn't add much but for frontends e.g. NWN-DTK by TPG it is a must to have some form of standardization so they can attach both user-friendly messages and also provide the option to have technical messages available.

>     * I am more concerned about the maintenance of this additional layer as this requires inputs from different teams. Now it is still easy to find each other during the active development phase, but it is hard to guarantee in the longer term. Also, any change request could introduce changes in different dependent components.

Definitely a concern and it seems this is also the core of Mark's worries on how to proceed here.

> For the time being, if we constantly receiving similar support requests from the users, I would look into the direction of having a Q&A section on the (architecture) documentation page that lists a few commonly seen technical/error messages to help users with the next steps.

Definitely good to add! Currently they are not receiving any technical/error messages so lets provide both some form of message and also a Q&A section.


Going to reply with another message shortly to alter the proposal to one that seems to have the most support currently.

**Comment by lfse-slafleur** [comment]
Currently it is a step too far to use codes for messages. Conversation with TPG has shown that some form of parsing will have to happen in the future, but there is no necessity currently. There is also not enough support from the development teams to pick it up now, so we are going to move forward with sending messages transparantly through OMOTES and keep the option open to use codes at a later point in time. The motivation to immediately move to codes was due to the expected investment needed by the optimizer and simulator teams but the current investigation has shown that the immediate investment is very little (~10ish messages expected). Therefore, the proposed plan for now:

- Datastructure will remain the same except the fields `dynamic_values` and `message_code` will not be used. We will continue to propose the message to be called `technical_message`.
- `asset_id` will remain optional so allow for asset-specific messages and global messages.
- Message is captured by the workers in whatever form the library underneath provides. In the case of mesido currently, we will capture specific exceptions and relay the message in those exceptions as error messages when an output ESDL cannot be calculated. At a later stage, mesido will also include warnings, info and debug messages if the ESDL is calculated successfully but there is more info that should be shared with a user.
- Frontends can optionally show the `technical_messages` provided by OMOTES. The requirement to be able to localize messages is not considered for now.
- We will provide a list of expected messages with their meaning in the form of a Q&A for both frontends and end-users. This is something frontends can optionally link to to provide end-users with more details regarding the error, warning or info message. Debug messages are considered to be seen only by developers of backend models e.g. mesido and simulator so these messages will be not be expected on the Q&A list. We will request the optimizer and simulator teams to maintain this list.

@MichielTukker  @MarkTNO @edwinmatthijssen  @KobusVanRooyen : Does this altered proposal have your support? If so, I will also check with the others before proceeding with a detailed design. Otherwise, please share your thoughts and proposed alterations.

**Comment by lfse-slafleur** [comment]
@MichielTukker @MarkTNO  @edwinmatthijssen  @KobusVanRooyen  I haven't had a reply yet on my Q if this altered proposal fits for you all. If so, we can start work on it at the cloud side.

**Comment by MarkTNO** [comment]
This sounds good!

**Comment by edwinmatthijssen** [comment]
To me too....!

**Comment by KobusVanRooyen** [comment]
@lfse-slafleur in principle it all sounds good, except the last bullet point about documentation. But I think we can discuss this item further and it will should not prevent code development work.

**Comment by lfse-slafleur** [comment]
@KobusVanRooyen Discussed through Teams. We will hold off on generating more documentation for each issue. Instead, we will see if the messages are info enough for users.

**Comment by lfse-slafleur** [comment]
First version is integrated in mapeditor.

**Comment by lfse-slafleur** [comment]
Simulator is also creating first messages based on heat limits.
